lms/views/dashboard/api/grading.py

```diff
@@ -44,43 +37,55 @@ def __init__(self, request) -> None:
-        assert assignment.lis_outcome_service_url, "Assignment without grading URL"
-        lti_registration = self.db.scalars(
-            select(LTIRegistration)
-            .join(ApplicationInstance)
-            .join(Grouping)
-            .where(Grouping.id == assignment.course_id)
-            .order_by(LTIRegistration.updated.desc())
-        ).first()
-        assert lti_registration, "No LTI registraion for LTI1.3 assignment"
 
-        sync_h_user_ids = [g["h_userid"] for g in self.request.parsed_params["grades"]]
+        if self.db.scalar(
+            select(GradingSync).where(
+                GradingSync.assignment_id == assignment.id,
+                GradingSync.status.in_(["scheduled", "in_progress"]),
+            )
+        ):
+            return 409
 
-        sync_lms_users = self.db.execute(
-            select(LMSUser.h_userid, LMSUser.lti_v13_user_id).where(
+        grading_sync = GradingSync(assignment_id=assignment.id, status="scheduled")
+        self.db.add(grading_sync)
+        self.db.flush()
+        sync_h_user_ids = [g["h_userid"] for g in self.request.parsed_params["grades"]]
+        sync_lms_users = self.request.db.execute(
+            select(LMSUser.h_userid, LMSUser.id).where(
                 LMSUser.h_userid.in_(sync_h_user_ids)
             )
         ).all()
         # Organize the data in a dict h_userid -> lti_user_id for easier access
         sync_lms_users_by_h_userid = {r[0]: r[1] for r in sync_lms_users}
-
-        grading_service = LTI13GradingService(
-            ltia_service=self.request.find_service(LTIAHTTPService),
-            line_item_url=None,
-            line_item_container_url=None,
-            product_family=None,  # type: ignore
-            misc_plugin=None,  # type: ignore
-            lti_registration=None,  # type: ignore
-        )
-        # Use the same timestamp for all grades of the same sync
-        grade_sync_time_stamp = datetime.now()
         for grade in self.request.parsed_params["grades"]:
-            grading_service.sync_grade(
-                lti_registration,
-                assignment.lis_outcome_service_url,
-                grade_sync_time_stamp,
-                sync_lms_users_by_h_userid[grade["h_userid"]],
-                grade["grade"],
+            self.db.add(
+                GradingSyncGrade(
+                    grading_sync_id=grading_sync.id,
+                    lms_user_id=sync_lms_users_by_h_userid[grade["h_userid"]],
+                    grade=grade["grade"],
+                )
             )
+        self.request.db.flush()
+        self.request.add_finished_callback(lambda _: sync_grades.apply_async(()))
+        return {}
+
+    @view_config(
+        route_name="api.dashboard.assignments.grading.sync",
+        request_method="GET",
+        renderer="json",
+        permission=Permissions.GRADE_ASSIGNMENT,
+    )
+    def get_auto_grading_sync(self):
```

Comment:
This is the /GET, it returns the last sync.

If it's a completed one, useful to show a "last_synced_at" or similar.

If in progress, useful to report progress.